### PR TITLE
CDPSDX-3898: Added new shape for DL that is efficient and can scale horizontally.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CDPSDX-3898

Introduced a new scale for a more robust medium duty DL shape, meant to be part of our effort for horizontal scaling.

List of improvements made:

1. Increase the memory allocation for services based on the inputs from service teams.
2. Capability to scale core hostgroup horizontally without any limitations.
3. Move master services that do not need scaling to a single host group.
4. Moved RAZ service to host group where CPU usage is not high.
5. Moved master services to a single host group that need not scale
6. Support Ranger user-sync and tag-sync HA.
7. Reduce the memory reserved for the cache.
8. Made sure the HBase master and region servers are in different host groups.
9. Capability to horizontal scale specific services by the empty host groups.
10. Decrease the non-required services memory usage, e.q: Solr Direct Memory
11. Enables scaling specific services in the data lake.

This shape will be private, i.e. internal-only until we GA it. This shape will also only be accessible in runtimes >= **7.2.17**.

### CODE REVIEW

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp
Comment (sxxgrc) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:7]: Based on the convo today, we only want this available on `7.2.17`

File: datalake/src/main/resources/duties/7.2.17/yarn/medium_duty_ha_scalable.json
Comment (sxxgrc) [reply datalake/src/main/resources/duties/7.2.17/yarn/medium_duty_ha_scalable.json:1]: This is more for completeness so that each shape is available on Yarn, but this is actually a duplicate of the `medium_duty_ha.json` and uses the same blueprint as normal HA, so this doesn't need to be tested.

File: datalake/src/main/resources/duties/7.2.17/aws/medium_duty_ha_scalable.json
Comment (kkalvagadda1) [reply datalake/src/main/resources/duties/7.2.17/aws/medium_duty_ha_scalable.json:70]: ok. Let's change the disk tye for master/gw and auxiliary host groups to Gp2 for AWS and equivalent for other cloud providers..

File: datalake/src/main/resources/duties/7.2.17/yarn/medium_duty_ha_scalable.json
Comment (kkalvagadda1) [reply datalake/src/main/resources/duties/7.2.17/yarn/medium_duty_ha_scalable.json:1]: Let's YARN as well. I'm good with the current approach.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:85]: All of these roles on the extra horizontally scalable hostgroups should be configured as different roles. Config parameters, otherwise, cannot be changed independently for these roles - and the same configs that would apply to the main instance of the role would apply to the scalable ones, and vice versa.

e.g hbase-REGIONSERVER-HS  would be a new role, which would be almost identical to hbase-REGIONSERVER-BASE - except with any custom memory, etc tuning that we may want to do based on this instance size.

File: datalake/src/main/resources/duties/7.2.17/aws/medium_duty_ha_scalable.json
Comment (sidseth) [inline datalake/src/main/resources/duties/7.2.17/aws/medium_duty_ha_scalable.json:188]: Please change this to gp2 as well (even though there's no attached nodes)

File: datalake/src/main/resources/duties/7.2.17/gcp/medium_duty_ha_scalable.json
Comment (sidseth) [inline datalake/src/main/resources/duties/7.2.17/gcp/medium_duty_ha_scalable.json:87]: Change all disks to ssds for the google template as well.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp
Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:85]: CM also has the option to set a config across roles - which can be used if a configuration needs to be changed across all roles.
These independent scalable roles should be controllable independently.

Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:85]: @sidseth  Why do we need to different configurations for the instances running on the extra horizontally scalable host groups?
Any examples?

Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:85]: One simple example: 
The RAZ hostGroup is sized to be larger, and accordingly we tune #threads or memory differently for these instances.

Similarly for the other services like Solr or HBase.

CM allows configs to be set at the service level - which can be used if everything is the same.
If however, we want the flexibility of changing configs only for the new host groups - that can be easily achieved in CM via hostGroup/role level configs.

A single role spanning hostGroups will always tie the configs together.

Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:85]: Make sense 
We will handle it as an improvement.  I don't want to hold this PR. We would like to get this shape merged so we can have functional and non-functional testing.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-medium-ha-scalable.bp:432]: 3 ?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:1001]: There a multiple places in the code that checks for medium duty. Shouldn't we review them to ensure the new scalable is covered in all those cases? Here is an example:

 public FlowIdentifier triggerRepairByCrn(String userCrn, String clusterCrn, SdxRepairRequest clusterRepairRequest) {
        SdxCluster cluster = sdxService.getByCrn(userCrn, clusterCrn);
        MDCBuilder.buildMdcContext(cluster);
        if (SdxClusterShape.MEDIUM_DUTY_HA.equals(cluster.getClusterShape()) && !entitlementService.haRepairEnabled(cluster.getAccountId())) {
            LOGGER.error("Cluster {} is Medium Duty and is not allowed to be repaired", cluster.getClusterName());
            throw new BadRequestException("Cannot repair Medium Duty cluster " + cluster.getClusterName() + " at this time");
        }
        return sdxReactorFlowManager.triggerSdxRepairFlow(cluster, clusterRepairRequest);
    }

**Review by kkalvagadda1 — CHANGES_REQUESTED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:1001]: Done as part of https://github.com/hortonworks/cloudbreak/pull/14312/commits/9cd74bfccb87817599fbf276a43011f362082590 please let me know if it looks good.

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:1001]: Done as part of https://github.com/hortonworks/cloudbreak/pull/14312/commits/9cd74bfccb87817599fbf276a43011f362082590 please let me know if it looks good.

Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:1001]: I left a comment. You want to separate this change from this PR?

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:1001]: I do not mind keeping this here as it is simple enough. As for your comment:

```
How about validateShape in SdxService.java
```

The line of code your original comment is pointing to addresses `validateShape` already.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: This needs to be a RAZ reference, correct?

Also - where is RAZ configured - injected via code? Which RG is it being placed on, and is it configuring the memory for the RAZ instance as well?

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:452]: kafka memory configuration required ?

**Review by kkalvagadda1 — APPROVED** [review]
The changes look good. 
We still need CLI and UI changes for it to be consumed.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:452]: Why is it not required ?

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: This hostGroup has "ranger-RANGER_ADMIN-BASE" - not RAZ.

There are NO references to RAZ in the template. How is it installed ?

>  in that case, the cm memory configuration will be enough
CM memory config will not run on a hostgroup which starts with 0 nodes.

Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: In medium-duty, RAZ is added by config providers in CB code and not by blueprint.
As RAZ is controlled by configuration it is not part of the blueprint.

we will need additional changes to it.

Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: Blueprint is static and changes controlled by configuration are done through config providers. 
That should be the case even for RAZ.

Comment (sidseth) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: Please make the changes in this PR itself (instead of a separate jira). We're trying to check this in with multiple different issues - while this is extremely close to being complete.

There's at least the following.
1. RAZ code changes to make sure it is on the razhg
2. Separating out roleConfigGroups for the new hostGroups - (this should be no more than a couple of hours)
3. kafka broker memory configuration not being set explicitly

Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp]: RAZ changes and new roleConfigGroups for the new hostGroups are done.

Memory configurations will be done separately.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:614]: Missing the safety valve properties

**Review by sidseth — COMMENTED** [review]
RAZ still needs to be fixed. I think this will be at: https://github.com/hortonworks/cloudbreak/blob/master/template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProvider.java#L37

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:51]: These were carried from the current medium-duty dl.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:62]: TagSync HA was available as of now. Should be available shortly.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:100]: Yes, it is missing. It should be added to RangerRazDatalakeConfigProvider.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:186]: Yes, this should be removed.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:436]: We had Hbase folks review the MD blueprint and there were no configuration changes recommended so we carried all the configurations in MD DL.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:526]: [CDPSDX-4005](https://jira.cloudera.com/browse/CDPSDX-4005)

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxUpgradeService.java
Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxUpgradeService.java:165]: When an OS upgrade is performed based on the shape a different order is picked. The order is defined in OrderedOSUpgradeSetRequest.  with [CDPSDX-3955](https://jira.cloudera.com/browse/CDPSDX-3955) OrderedOSUpgradeRequestProvider will update to provide the right order for the scalable data lake.

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:409]: Yes, is possible that there may not be consumers of this service. we are not sure of it. we can evaluate it and remove it later.

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProvider.java
Comment (lajosrodek) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProvider.java:1]: Please cover with UT.

File: datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterShape.java
Comment (lajosrodek) [inline datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxClusterShape.java:20]: Please cover with UT.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
A few comments, mostly around missing tests. Would be nice to get some feedback from Raven folks as well.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxRepairService.java
Comment (horadla23) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxRepairService.java:1]: +1

File: template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProvider.java
Comment (horadla23) [inline template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProvider.java:38]: please define a static string list in the class and check if that list contains `hg.getName`

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java
Comment (horadla23) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:180]: this message will be printed in case of MEDIUM_DUTY  too as I see, thus this should be fixed

**Review by horadla23 — CHANGES_REQUESTED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java
Comment (horadla23) [reply datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:180]: or will we recommend scalable instead of medium duty every time?

**Comment by horadla23** [comment]
We would also need e2e tests for this (provision, light -> scalable resize, etc.). Do you have any task for this? We should not enable this for customers if we do not have tests for this feature

**Comment by horadla23** [comment]
Also I would suggest to introduce an entitlement for this to prevent accidental usage of this shape if it's gets released.

File: datalake/src/main/resources/duties/7.2.17/yarn/scalable.json
Comment (sxxgrc) [reply datalake/src/main/resources/duties/7.2.17/yarn/scalable.json:3]: Yes this is intentional albeit a little weird, this is only for yarn and for this change we will not be supporting the new shape for yarn setups. I simply added this here as a placeholder and it functions identically to the medium duty setup.

In the future we will add in the yarn version of this shape.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:180]: I think for now we will keep this as medium duty, but in the future we can change this to scalable as that will be our ideal shape for everyone.
As for your question about upgrade/resize. Upgrade should function normally within this shape, and resize is a very simply change for this but is not part of this change. Upgrade has not been tested so we cannot confirm its functionality just yet.

**Comment by sxxgrc** [comment]
> We would also need e2e tests for this (provision, light -> scalable resize, etc.). Do you have any task for this? We should not enable this for customers if we do not have tests for this feature

Yes this will be a part of https://jira.cloudera.com/browse/CDPSDX-3985.

**Comment by sxxgrc** [comment]
> Also I would suggest to introduce an entitlement for this to prevent accidental usage of this shape if it's gets released.

This is something we've considered and is still in discussion, I will make sure to arrive to a consensus before trying to get this merged. Our original idea was to put this out as a tech preview only on 7.2.17 which will be released in a few months, so we thought it would be pretty safe, but I agree that an entitlement will ensure avoiding any potential customer issues.

**Comment by kkalvagadda1** [comment]
> @lajosrodek @horadla23 This shape is only available for runtime 7.2.17 so this shape will not be available for customers.

We are considering introducing an entitlement for this new shape but it will be a new PR.

**Comment by sxxgrc** [comment]
@horadla23 @lajosrodek I have added unit tests and the additional changes you requested in the new commit, feel free to take a look. There were also some comments I responded to directly.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java]: @kkalvagadda1 that does not apply to this case, we are recommending a specific cluster shape here.

File: datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java
Comment (sidseth) [inline datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/SdxRuntimeUpgradeService.java:180]: Mostly a nit... 

This message still doesn't make sense to me. I know an overall fix to define the OS RU for the Scalable shape will go in separately.

Until then: 
1) The code allows an RU for the scalable shape
2) Logs that the shape must be MEDIUM DUTY to perform an RU.

File: datalake/src/main/resources/duties/7.2.17/azure/scalable.json
Comment (sidseth) [inline datalake/src/main/resources/duties/7.2.17/azure/scalable.json:116]: v24 ?

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (sidseth) [inline core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:99]: Good to know, that this works without any servicces being configured here (maybe because the service gets added via the RazProvider in code).

Will need to test: What happens to this HostGroup is RAZ is not enabled.

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProviderTest.java
Comment (lajosrodek) [inline template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProviderTest.java:200]: Is the expectation to have RAZ in both `master` and `razhg`? If so, I suggest also checking for `master`. (I know `getAdditionalServicesWhenRazIsEnabledWithRightCm())` kinda covers this.)

File: datalake-api/src/test/java/com/sequenceiq/sdx/api/model/SdxClusterShapeTest.java
Comment (lajosrodek) [inline datalake-api/src/test/java/com/sequenceiq/sdx/api/model/SdxClusterShapeTest.java:34]: Please extract as a constant.

File: datalake-api/src/test/java/com/sequenceiq/sdx/api/model/SdxClusterShapeTest.java
Comment (lajosrodek) [inline datalake-api/src/test/java/com/sequenceiq/sdx/api/model/SdxClusterShapeTest.java:22]: Please extract as a constant.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Thanks for the fixes. A few minor additional asks.

**Comment by horadla23** [comment]
Is there any decision about the entitlement we talked above?

File: core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp
Comment (kkalvagadda1) [reply core/src/main/resources/defaults/blueprints/7.2.17/cdp-sdx-scalable.bp:99]: This was tested.
When RAZ is disabled, provisioning was successful. There will be a failure if we try to scale it.

**Comment by kkalvagadda1** [comment]
> Is there any decision about the entitlement we talked above?

@horadla23 There will be an entitlement for scalable data lake. We prefer it to be a separate PR right after this PR is merged. What do you say?

File: template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProviderTest.java
Comment (sxxgrc) [reply template-manager-cmtemplate/src/test/java/com/sequenceiq/cloudbreak/cmtemplate/configproviders/raz/RangerRazDatalakeConfigProviderTest.java:200]: Yes they will both be present I will check for both in this test.

**Comment by sxxgrc** [comment]
I have addressed the most recent comments in the latest commit: 4802585aac2c514d35755979a2d7c013df8213d2

Once this has been approved we can squash and merge.

**Review by sidseth — APPROVED** [review]

**Review by lajosrodek — APPROVED** [review]
Thanks, LGTM. Could you please squash the changes into a single commit? (As I see, they all are related to CDPSDX-3898 and logically belong together.)

**Review by lacikaaa — CHANGES_REQUESTED** [review]
to prevent accidental merge: squash commits

**Comment by kkalvagadda1** [comment]
> Thanks, LGTM. Could you please squash the changes into a single commit? (As I see, they all are related to CDPSDX-3898 and logically belong together.)

That was done on purpose to make it easy for the reviewers. We will be squashing the commits.

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**
